index.bs

```diff
@@ -2916,6 +2916,214 @@ value and terminate the operation.
 
 </div>
 
+### Signal Credential Changes to the Authenticator - PublicKeyCredential's `signal()` Method {#sctn-signal}
+
+<xmp class="idl">
+    partial interface PublicKeyCredential {
+        static Promise<undefined> signal(PublicKeyCredentialSignalOptions options);
+    };
+
+    dictionary PublicKeyCredentialSignalOptions {
+        USVString                              rpId;
+        Base64URLString                        unknownCredentialId;
+        PublicKeyCurrentCredentialDetails      currentCredentials;
+    };
+
+    dictionary PublicKeyCurrentCredentialDetails {
+        required Base64URLString               userId;
+        PublicKeyCurrentCredentialUserDetails  user;
+        sequence<Base64URLString>              allAcceptedCredentialIds;
+    };
+
+    dictionary PublicKeyCurrentCredentialUserDetails {
+        required DOMString                     name;
+        required DOMString                     displayName;
+    };
+</xmp>
+
+[=[WRPS]=] may use {{PublicKeyCredential/signal()}} to signal [=authenticators=]
+the state of [=public key credentials=], so that incorrect or revoked
+credentials may be updated or removed. [=User agents=] provide this
+functionality opportunistically, since an authenticator may not support updating
+its [=credentials map=] or it may not be attached at the time the request is
+made. Furthermore, in order to avoid revealing information about a user's
+credentials without [=user consent=], the {{PublicKeyCredential/signal()}}
+operation does indicate whether the operation succeeded. A successfully resolved
+promise only indicates that the {{PublicKeyCredentialSignalOptions}} object was
+well formed and accepted.
+
+A number of <dfn>credential reports</dfn> can be signalled by [=[WRPS]=]. Each
+[=credential report=] declares a <dfn>validation algorithm</dfn> and <dfn>
+authenticator actions</dfn>. The input for both these algorithms is the
+{{PublicKeyCredentialSignalOptions}} |options|.
+
+#### <dfn>unknownCredentialId</dfn> report #### {#sctn-unknownCredentialIdReport}
+
+Signals that a [=credential id=] was not recognized by the [=[WRP]=], e.g.
+because it was deleted by the user.
+
+The [=unknownCredentialId=] [=validation algorithm=] is as follows:
+1. If the result of [=base64url encoding | base64url decoding=]
+    <code>|options|.{{PublicKeyCredentialSignalOptions/unknownCredentialId}}</code>
+    is an error, then throw a {{TypeError}}.
+    
```

Comment:
Do we need anything here, in the name of avoiding abuse of this call (either by RP or malicious JS), to have the client remember the credential ID in the last resolved `.get()` response and only allow the RP to specify that ID in a subsequent `unknownCredentialId` report? The idea being, if an RP receives a response with an `id` of "credA", they shouldn't ever set anything but `"credA"` in their unknown credential Report.

Thinking one step further, I'd say if that idea has merit then what's the point in the RP specifying a credential ID? The RP could more simply signal "I didn't recognize that credential" and the client uses its own knowledge of the last used credential's ID to deprecate that credential. No value validation needs to go on, the RP passes a simpler value for this specific report... 🤔

Reply:
In terms of abuse, malicious JS can do way more nasty things, such as steal assertions and intercept credential creation requests that we can't meaningfully defend against so I would consider it outside the threat model. I lack the imagination to see what a malicious RP could do with this endpoint as specified, do you have an example?

> The RP could more simply signal "I didn't recognize that credential" and the client uses its own knowledge of the last used credential's ID to deprecate that credential. No value validation needs to go on, the RP passes a simpler value for this specific report... 🤔

I think this is conceptually fine, but significantly more complicated in terms of implementation for no discernible benefit. We'd have to specify how long we remember the credential id for, whether we do it across navigations, and what the remembered credential id is keyed to (the frame? the global browser object?). I would prefer not to go this route unless we find a good reason to avoid RPs from specifying arbitrary credential ids.

Reply:
> We'd have to specify how long we remember the credential id for, whether we do it across navigations, and what the remembered credential id is keyed to

Maybe it could be tied by exposing the signal as a callback method exposed on the `PublicKeyCredential` instance, but I agree this seems overcomplicated for little benefit.